### PR TITLE
Added tab keybinds/refactor to remove inline functions that shouldn't be inline and also rename basically every var in index

### LINKED ISSUE #23: Keyboard shortcuts for tab support feature.
Thank you for supporting tab.
Please add keyboard shortcuts:

| Command                     | Shortcut                       |
|-----------------------------|--------------------------------|
| **Go one Tab to the Right** | **`Ctrl` + `Tab`**             |
| **Go one Tab to the Left**  | **`Ctrl` + `Shift` + `Tab`**   |
| **New Tab**                 | **`Ctrl` + `T`**               |
| **Close Tab**               | **`Ctrl` + `W`**               |
| **Close Tab**               | **`Ctrl` + `F4`**              |
| Go one Tab to the Right     | `Ctrl` + `Page Down`           |
| Go one Tab to the Left      | `Ctrl` + `Page Up`             |
| Go to Tab 1                 | `Ctrl` + `1`                   |
| Go to Tab 2                 | `Ctrl` + `2`                   |
| Go to Tab 3                 | `Ctrl` + `3`                   |
| Go to Tab 4                 | `Ctrl` + `4`                   |
| Go to Tab 5                 | `Ctrl` + `5`                   |
| Go to Tab 6                 | `Ctrl` + `6`                   |
| Go to Tab 7                 | `Ctrl` + `7`                   |
| Go to Tab 8                 | `Ctrl` + `8`                   |
| Go to Last Tab              | `Ctrl` + `9`                   |
| Undo Close Tab              | `Ctrl` + `Shift` + `T`         |
| Move Tab Left               | `Ctrl` + `Shift` + `Page Up`   |
| Move Tab Right              | `Ctrl` + `Shift` + `Page Down` |
| Move Tab to start           | `Ctrl` + `Shift` + `Home`      |
| Move Tab to end             | `Ctrl` + `Shift` + `End`       |

Thanks.

### LINKED ISSUE #28: Dark scrollbar like Firefox
Please use the dark scrollbar like Firefox.

# NightPDF
image

# Firefox
image

Thanks.

### LINKED ISSUE #33: [Linux] [Flathub] Drag and drop does not work!
Hi, Drag and Drop does not work!
https://www.dropbox.com/s/soogh7h27wc2j91/2023-06-11%2001-57-07.mp4?dl=1

### LINKED ISSUE #35: Add "About" item
Please add the "About" item to the Help menu.

image

Now I can't see the installed version.

This feature is on all apps for example VS Code.

image

image

### LINKED ISSUE #36: Fix tabs titles
There is a small bug in the tabs titles.

The file name is "03. DevOps.pdf"

image

But in tab displayed: "DevOps.pdf .03"

image

### LINKED ISSUE #38: [WIndows] Support multi file drag
Please support multi-file drag on Windows.

image

Now only open a file.
I expect open all files to separate tabs.

### PR DESCRIPTION
**Pull request by zeyus**
Fixes #23 
Fixes #33
Fixes #28
Fixes #38
Fixes #35
Fixes #36

This is an implementation of tab keyboard navigation.

- Navigation to previous, next, index (1-8) and 9 (last tab)
- Tab moving to previous, next, first, and last. 
- Tab open
- Tab close
- Reopen closed tabs

New settings page:

- Customizable keybinds
- About page
- Other settings (WIP)

This also fixes:

- an issue with PDF.js capturing the `ctrl+o` and `ctrl+p` keybinds. Now they are handled by NightPDF instead :)
- an issue where the DevTools window would intermittently not open.
- an issue when opening a PDF that was already open would not switch to the existing tab with that PDF
- PDF.js scrollbar styles (on my windows PC, I have dark mode, and for some reason the scrollbar was still light / grey)
- devtools within PDF.js (when `DEBUG=true`
- Multi-file opening
- Multi-file drag

-- message from luna ---
This basically rewrites NightPDF in many areas(read this is basically a 3.0 and cleans up a large portion of technical debt brought from the javascript version). while i would like to immediately to cut a release when this is done, I do think it would be wiser to release a beta or 2 to ensure there are no errors.

### CODE REVIEW

**Comment by Lunarequest** [comment]
I'll need a bit to review this. I have noticed!

**Comment by zeyus** [comment]
> I'll need a bit to review this. I have noticed!

No stress from my end!

**Comment by Lunarequest** [comment]
I'm working on adding configurability to your change I've introduced `electron-store`  as a dep it is a simple key value store.

TODO: expose settings on the UI to change keybindings

**Comment by zeyus** [comment]
> I'm working on adding configurability to your change I've introduced `electron-store` as a dep it is a simple key value store.
> 
> TODO: expose settings on the UI to change keybindings

Sounds great. Let me know if you want me to look at something :) 

I'd also be interested in discussing an incremental refactor at some point, to make the code more modular and organized.

**Comment by Lunarequest** [comment]
Would you be able to add the settings ui, I'm not able to come up with a good ui

**Comment by zeyus** [comment]
> Would you be able to add the settings ui, I'm not able to come up with a good ui

Def, I'll take a look this evening and see where I get :)

**Comment by Lunarequest** [comment]
I noticed what you were doing, so anything in render needs to be bundled since there is no access to the node api i.e require doesn't work. If you need access to the node api you need to go through the preload(which is also sandboxed so only electron remote module, and node's events, timers, and url modules. to access more you use `ipcRenderer.invoke/send`

**Comment by Lunarequest** [comment]
made some minor tweaks so helpers are now turned to js, I've added a version field to the settings scheme in case in the future we want to change parameter names or come up with other ideas that use  electron-store we should be able to migrate without any issues. I've also noticed keybinds has no default value which results in errors on start up

**Comment by zeyus** [comment]
> made some minor tweaks so helpers are now turned to js, I've added a version field to the settings scheme in case in the future we want to change parameter names or come up with other ideas that use electron-store we should be able to migrate without any issues. I've also noticed keybinds has no default value which results in errors on start up

Awesome, thanks I will do some more today, I just have some meetings and uni work first.

About the defaults, when I created the store schema, initially I had a seperate "defaults" object which worked (I only had two so far) and I worked perfectly, afterwards, I moved the defaults into the schema itself, I had not tested that yet because of the issues you mentioned with the JS (though I thought you hluldtbe able to bundle your own modules but I'm still not to familiar with this process)...if the defaults do not get loaded from the schema `default` key, that's fine, I will just create a seperate object again with all the defaults (also makes it easy to "reset key binds" or "reset all settings" :)

**Comment by zeyus** [comment]
so yes, it appears that schema defaults get ignored, but you can send in an object with the defaults instead and that works.

for the UI I have an idea of how to approach it, most simply, a menu / list on the left with settings categories, an on the right a pane with the settings. for keybinds i was thinking of replicating a similar vibe to Tidal, but with a single column instead of two, and allowing potentially two keybinds per shortcut (e.g. ctrl+w and ctrl+f4).

image

in terms of implementation, it should have a delete keybind button for each individual bind, a reset all button to set to defaults, and then a click to update/add keybind, and then a div that appears that captures all keypresses except for escape which would cancel the bind.

also i think it might be worth making a keymap to automate some of the UI, so it can be used in the settings but also potentially in a help window or other places, e.g.

```js
{
  CloseTab: {
    description: "Close the currently active tab",
    default: default_settings["keybinds"]["CloseTab"],
    weight: 0, // possibly, a way to order in the settings or other pages rather than alpha
  },
  // ...
}
```

This may also make internationalization easier if that is something that might be considered in the future, reusing strings rather than including them in the actual page / render.

**Comment by Lunarequest** [comment]
hi would you happen to be on discord? I think it would be simpler to talk over there vs here

**Comment by zeyus** [comment]
> hi would you happen to be on discord? I think it would be simpler to talk over there vs here

yes! `zeyus#9721` feel free to ping me whenever, I will not always be able to reply straight away, the next few hours I'm moving all over the place, but I'll be at my PC again this afternoon for a little while, then later on this evening.

**Comment by Lunarequest** [comment]
given this is more or less a complete rewrite of NightPDF, should we consider dropping, AppImage. From what I've heard AppImage isn't truly build once run everywhere, non-sandboxed and actually was broken on ubuntu and other major distros due to the creators refusal to use fuse3. if people need support for a distro we don't build packages for the flatpak still exists

**Comment by zeyus** [comment]
> given this is more or less a complete rewrite of NightPDF, should we consider dropping, AppImage. 

Makes sense to me, I don't know how common it is but I've been using tons of different Linux flavors over the years and never onced used an AppImage. Only makes sense to support it if people use it and there's some benefit.

> if people need support for a distro we don't build packages for the flatpak still exists

Agreed, btw where is the flatpak being built / configured? What I wanted to check as well is to see that the file associations were correct, but I haven't built a flapak before and I tried manually out of curiosity and it was a pain in the ass using a Debian instance I spun up 🫠

**Comment by Lunarequest** [comment]
> Agreed, btw where is the flatpak being built / configured? What I wanted to check as well is to see that the file associations were correct, but I haven't built a flapak before and I tried manually out of curiosity and it was a pain in the ass using a Debian instance I spun up 🫠

flatpak is built here https://github.com/flathub/io.github.lunarequest.nightpdf, its quite easy to build a flatpak if you use flatpak-builder. in this case we have to use a few tools to generate sources for node modules

File: app/render/settings.ts [hunk 1]
```diff
@@ -273,7 +275,21 @@ async function nightPDFSettings() {
 		);
 
 		// @todo: add event listener to save button, check for duplicates
-		// ...
+		saveButton?.addEventListener(
+			"click",
+			() => {
+				const values = document.getElementById(
```
Comment (zeyus) [inline app/render/settings.ts:309]: I know this is WIP, but you can use `newKeybind`. I'll see if I have some time tonight to refactor `type Keybind` into a class

**Comment by zeyus** [comment]
Last commit failed because I tried using github.dev on mobile and it's just not worth it.

**Review by Lunarequest — COMMENTED** [review]
:tada:

**Comment by zeyus** [comment]
@Lunarequest 

settings page

image

keybind overlay:

image

**Comment by Lunarequest** [comment]
should we drop the general settings tab? there's nothing i can find we can add there

**Comment by zeyus** [comment]
> should we drop the general settings tab? there's nothing i can find we can add there

We can, I was hoping to put the option to "auto maximize on open" so it can be turned off :) but it's not implemented yet so it can be dropped for now.

**Comment by Lunarequest** [comment]
> > should we drop the general settings tab? there's nothing i can find we can add there
> 
> We can, I was hoping to put the option to "auto maximize on open" so it can be turned off :) but it's not implemented yet so it can be dropped for now.

if you do intend to implement it in this pr then sure we can keep it, otherwise i vote for dropping it

**Comment by zeyus** [comment]
> if you do intend to implement it in this pr then sure we can keep it, otherwise i vote for dropping it

Sure, let me see if I can get it done today :)

**Comment by zeyus** [comment]
> > > should we drop the general settings tab? there's nothing i can find we can add there
> > 
> > 
> > We can, I was hoping to put the option to "auto maximize on open" so it can be turned off :) but it's not implemented yet so it can be dropped for now.
> 
> if you do intend to implement it in this pr then sure we can keep it, otherwise i vote for dropping it

I've updated the PR with the setting for maximizing the window on open :)

**Comment by Lunarequest** [comment]
i'm going to look into getting a new icon for nightpdf,  if I'm able to before this pr is done I'll add it as part of this pr otherwise I'll hold of doing a full release and publishing a beta till i can get it

**Comment by zeyus** [comment]
> i'm going to look into getting a new icon for nightpdf, if I'm able to before this pr is done I'll add it as part of this pr otherwise I'll hold of doing a full release and publishing a beta till i can get it

Sounds great, are you going to make it? Otherwise I could give it a try, if there's a specific style / design idea you have.

**Comment by Lunarequest** [comment]
> > i'm going to look into getting a new icon for nightpdf, if I'm able to before this pr is done I'll add it as part of this pr otherwise I'll hold of doing a full release and publishing a beta till i can get it
> 
> Sounds great, are you going to make it? Otherwise I could give it a try, if there's a specific style / design idea you have.

I'm actually looking to get it commissioned with brand guidelines n stuff

**Comment by Lunarequest** [comment]
btw would you say this pr is complete? I'll merge it and release a beta and we can release a full one after getting the new icon

**Comment by zeyus** [comment]
> btw would you say this pr is complete? I'll merge it and release a beta and we can release a full one after getting the new icon

pretty much, I'm working right now on implementing the clear keybind so you can delete them and add new ones. i'm pretty much there. also I'm not sure how deep you want to go into CSS and style stuff right now if you're getting a style guide but the different panels in the settings are not really consistent in display, but that's a minor non-functional problem

**Comment by zeyus** [comment]
> btw would you say this pr is complete? I'll merge it and release a beta and we can release a full one after getting the new icon

and done :) yeah i think we can go ahead. I think the code got a bit spaghetti-like in the end, but i think it's fine for now, it would be nice to organize it in a bit more structured way though at some point (but that means designing an overall architecture, like how we would deal with all types of settings, and pages etc, because the bind popup, and the clear bind etc essentially could be api endpoints, or some kind of MVC style thing)

**Comment by xmha97** [comment]
@Lunarequest Hi, keyboard shortcuts don't work :(
I'm using Windows 11

**Comment by Lunarequest** [comment]
Honestly at this point beyond builds I can't guarantee anything for windows. I don't have any windows machines and spinning up a vm just for this is more work then I'm interested in

**Pull request merged**